File: Phase-1/Sorting/HeapSort/heap.cpp
#include <iostream>
#include <vector>
using namespace std;

class Heap {
    public:
    int arr[100];
    int size;

    // Constructor
    Heap(){
        arr[0] = -1;
        size = 0;
    }

    // Insertion in Max Heap
    void insert(int val) {
        size = size + 1;
        int index = size;
        arr[index] = val;

        while (index > 1) {
            int parent = index / 2;

            if (arr[parent] < arr[index]) {
                swap(arr[parent], arr[index]);
                index = parent;
            } else {
                return;
            }
        }
    }

    // Delete from Heap
    void deleteFromHeap() {
        if (size == 0) {
            cout << "Nothing to delete" << endl;
            return;
        }

        // Step 1: Put last element in root
        arr[1] = arr[size];
        size--;

        // Step 2: Take root to its correct position
        int i = 1;
        while (i < size) {
            int leftIndex = 2 * i;
            int rightIndex = 2 * i + 1;

            if (leftIndex <= size && arr[i] < arr[leftIndex]) {
                swap(arr[i], arr[leftIndex]);
                i = leftIndex;
            } else if (rightIndex <= size && arr[i] < arr[rightIndex]) {
                swap(arr[i], arr[rightIndex]);
                i = rightIndex;
            } else {
                return;
            }
        }
    }

    // Print Heap
    void print() {
        for (int i = 1; i <= size; i++) {
            cout << arr[i] << " ";
        }
        cout << endl;
    }
};


void heapify(int arr[], int n, int i) {
    int largest = i; 
    int left = 2 * i; 
    int right = 2 * i + 1; 

    if (left < n && arr[left] > arr[largest])
        largest = left;

    if (right < n && arr[right] > arr[largest])
        largest = right;

    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

int main() {
    Heap h;

    vector<int> vec = {54, 53, 55, 52, 50};

    for (int i = 0; i < vec.size(); i++) {
        h.insert(vec[i]);
    }

    h.print();

    h.deleteFromHeap();
    h.print();

    int arr[] = {-1, 54, 53, 55, 52, 50};
    int n = sizeof(arr) / sizeof(arr[0]);
    for (int i = n/2; i >= 1; i--)
    {
        heapify(arr, n, i);
    }

    cout << "Array representation of Heap after heapify: ";
    for (int i = 1; i < n; i++) {
        cout << arr[i] << " ";
    }cout << endl;
    

    return 0;
}

File: Phase-1/Sorting/HeapSort/heap.py
class Heap: 
    # Insert element into the heap
    def insert(self,value): 
        # Rebuild heap to ensure valid heap state before inserting
        self.buildHeap()
        
        index = self.size
        self.arr.append(value)
        self.size += 1

        while(index > 0):
            parent = (index-1)//2
            if (self.arr[parent]<self.arr[index]):
                self.arr[parent],self.arr[index] = self.arr[index],self.arr[parent]
                index = parent
            else:
                return

    
    # Heapify the subtree rooted at target index
    def heapify(self,target, heapSize):
        largest = target
        left = 2 * target +1
        right= 2 * target +2

        if (left< heapSize and self.arr[left]>self.arr[largest]):
            largest = left
        if (right< heapSize and self.arr[right]>self.arr[largest]):
            largest = right

        if (largest != target):
            self.arr[target],self.arr[largest] = self.arr[largest],self.arr[target]
            self.heapify(largest, heapSize)


    # Extract the maximum element from the heap
    def extract(self):
        if (self.size == 0):
            return

        self.arr[0],self.arr[self.size -1] = self.arr[self.size-1], self.arr[0]
        self.size -= 1

        self.heapify(0,self.size)


    # Perform heap sort on the array
    def heapSort(self):
        n = self.size

        while (n>1):
            self.arr[0],self.arr[n-1] = self.arr[n-1],self.arr[0]
            n -= 1
            self.heapify(0, n)
    

    # Build the heap from the current array
    # range(start, stop, step)
    def buildHeap(self):
        for i in range((self.size -1)//2, -1, -1):
            self.heapify(i, self.size)
        

    # Build the heap from the current array
    def __init__(self, items=[], ):
        self.arr = items
        self.size = len(self.arr)
        if len(self.arr) > 0:
            self.buildHeap()


    # Print the current state of the heap
    def printHeap(self):
        printArr = []
        for i in range(self.size):
            printArr.append(self.arr[i])
        print(printArr)
       

array = [50,70,40,10,20, 100]
h = Heap(array)
print("initial heap")
h.printHeap()

h.heapSort()
print("after Heap Sort")
h.printHeap()

h.insert(80)
print("after inserting 80")
h.printHeap()

h.extract()
print("after extracting max")
h.printHeap()

print("Heap Size: ", h.size)

